### PR TITLE
Add promulgation to audit log.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by fabricematrat** [comment]
QA:

check the promulgation is written in audit log.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1055/
Test PASSed.

File: audit/audit.go [hunk 1]
```diff
@@ -15,6 +15,9 @@ const (
 	// OpSetPerm represents the setting of ACLs on an entity.
 	// Required fields: Entity, ACL
 	OpSetPerm Operation = "set-perm"
+	// OpSetPromulgation represents the promulgation on an entity.
```
Comment (mhilton) [inline audit/audit.go]: Generally there would be a blank line here.

File: audit/audit.go [hunk 2]
```diff
@@ -25,9 +28,10 @@ type ACL struct {
...
-	User   string           `json:"user"`
-	Op     Operation        `json:"op"`
-	Entity *charm.Reference `json:"entity,omitempty"`
-	ACL    *ACL             `json:"acl,omitempty"`
+	Time         time.Time        `json:"time"`
+	User         string           `json:"user"`
+	Op           Operation        `json:"op"`
+	Entity       *charm.Reference `json:"entity,omitempty"`
+	ACL          *ACL             `json:"acl,omitempty"`
+	Promulgation bool             `json:"promulgation,omitempty"`
```
Comment (mhilton) [inline audit/audit.go]: Promulgation, or promulgated?

File: internal/v4/api.go [hunk 3]
```diff
@@ -978,7 +957,7 @@ func (h *ReqHandler) serveDelegatableMacaroon(_ http.Header, req *http.Request)
 	return m, nil
 }
 
-// GET id/promulgate
+// PUT id/promulgate
```
Comment (mhilton) [inline internal/v4/api.go:978]: thanks

File: internal/v4/archive_test.go [hunk 4]
```diff
@@ -283,7 +283,7 @@ func (s *ArchiveSuite) TestConcurrentUploads(c *gc.C) {
 	// by closing errorBodies.
 	try := make(chan struct{})
 	go func(try chan struct{}) {
-		for _ = range try {
+		for range try {
```
Comment (mhilton) [inline internal/v4/archive_test.go]: I got compilation errors from this before, but jenkins seems happy.

**Comment by mhilton** [comment]
:+1: LGTM with a couple of comments.

File: internal/v4/archive_test.go (hunk 4 above)
Comment (fabricematrat) [inline internal/v4/archive_test.go]: @mhilton make simplify does that, may be it's 1.4, but Jenkins seems to be 1.2 so let's track when you have the compilation error again.

File: audit/audit.go (hunk 1 above)
Comment (fabricematrat) [inline audit/audit.go]: Done.

File: audit/audit.go (hunk 2 above)
Comment (fabricematrat) [inline audit/audit.go]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1056/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1057/
Test PASSed.

File: internal/v4/api.go [hunk 5]
```diff
@@ -1033,3 +1025,27 @@ func (h *ReqHandler) resolveId(f resolvedIdHandler) router.IdHandler {
...
+		Op:     audit.OpSetPerm,
+		Entity: &id.URL,
+		ACL: &audit.ACL{
+			Read:  read,
+			Write: write,
+		},
+	})
+}
+
+func (h *ReqHandler) auditSetUser(e *audit.Entry) *audit.Entry {
```
Comment (rogpeppe) [inline internal/v4/api.go]: The name of this is of the same form as auditSetPerms, but they do very different things - one returns a new audit entry that records a "set perms" operation; the other sets the user in an audit entry.

How about instead defining a method on ReqHandler to add an audit entry?

```
func (h *ReqHandler) addAudit(e *audit.Entry) {
    if h.auth.Username == "" && !h.auth.Admin {
        panic("No auth set in ReqHandler")
    }
    e.User = h.auth.Username
    if h.auth.Admin && e.User == "" {
        e.User = "admin"
    }
    h.Store.AddAudit(e)
    if testAddAuditCallback != nil {
        testAddAuditCallback(*e)
    }
}
```

Then everywhere that adds an audit entry would use that method
and automatically gain the benefit of the user being added.

File: internal/v4/api_test.go [hunk 6]
```diff
@@ -2535,6 +2553,23 @@ func (s *APISuite) TestPromulgate(c *gc.C) {
...
+			c.Assert(calledEntities, jc.DeepEquals, []audit.Entry{{
+				User:        test.expectUser,
+				Op:          audit.OpSetPromulgation,
+				Entity:      ref,
+				Promulgated: test.expectPromulgate,
+			}})
+		} else {
+			c.Assert(len(calledEntities), gc.Equals, 0)
+		}
+		calledEntities = []audit.Entry{}
```
Comment (rogpeppe) [inline internal/v4/api_test.go]: s/[]audit.Entry{}/nil/

File: internal/v4/api_test.go [hunk 7]
```diff
@@ -2216,6 +2218,8 @@ var promulgateTests = []struct {
 	expectBaseEntities: []*mongodoc.BaseEntity{
 		storetesting.NewBaseEntity("~charmers/wordpress").Build(),
 	},
+	expectPromulgate: false,
```
Comment (rogpeppe) [inline internal/v4/api_test.go]: I'd just leave out this field whenever it's false.

File: audit/audit.go [hunk 8]
```diff
@@ -15,6 +15,10 @@ const (
 	// OpSetPerm represents the setting of ACLs on an entity.
 	// Required fields: Entity, ACL
 	OpSetPerm Operation = "set-perm"
+
+	// OpSetPromulgation represents the promulgation on an entity.
+	// Required fields: Entity, Promulgation
+	OpSetPromulgation Operation = "set-promulgation"
```
Comment (rogpeppe) [inline audit/audit.go]: "set-promulgated" (and OpSetPromulgated) would seem to fit better with the existing names in the code and API.

**Comment by rogpeppe** [comment]
This looks great, thanks. LGTM with the above fixes applied.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1058/
Test PASSed.

File: internal/v4/api.go (hunk 5 above)
Comment (fabricematrat) [inline internal/v4/api.go]: Done.

File: internal/v4/api_test.go (hunk 6 above)
Comment (fabricematrat) [inline internal/v4/api_test.go]: Done.

File: internal/v4/api_test.go (hunk 7 above)
Comment (fabricematrat) [inline internal/v4/api_test.go]: Done.

File: audit/audit.go (hunk 8 above)
Comment (fabricematrat) [inline audit/audit.go]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1059/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1060/
Test PASSed.

File: internal/v4/api.go (hunk 3 above)
Comment (fabricematrat) [inline internal/v4/api.go:978]: you're welcome.

**Comment by fabricematrat** [comment]
:shipit: thanks @mhilton  && @rogpeppe

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**